projects/snapshot_bot/memory.cc

```diff
@@ -92,7 +92,11 @@ void MemoryBase::trainRoute(const Navigation::ImageDatabase &route,
                 backgroundEx->check();
             }
 
-            train(snapshot.first, snapshot.second);
+            // Process snapshot
+            // **NOTE** ImageInput classes are NOT THREAD-SAFE so this needs to be done in serial
+            const auto processedSnapshot = imageInput.processSnapshot(snapshot);
+
+            train(processedSnapshot.first, processedSnapshot.second);
```

Comment:
If you're using C++17 now, this could be:

```suggestion
            const auto [snapshot, mask] = imageInput.processSnapshot(snapshot);

            train(snapshot, mask);
```